base/src/main/java/org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java

```diff
@@ -1274,7 +1274,7 @@ public SSLEngineResult unwrap(ByteBuffer src, ByteBuffer[] dsts, int offset, int
             SSLException excpt = ssl_exception;
             ssl_exception = null;
             handshake_state = SSLEngineResult.HandshakeStatus.NEED_WRAP;
-            tryCleanup();
+            cleanup();
```

Comment:
Wow, this is a really interesting change.

I think, `finalize` made sense to me, as a catch-all for any failures of applications to conform with expectations. [If my memory serves correctly, `Cleaner`](https://docs.oracle.com/javase/9/docs/api/java/lang/ref/Cleaner.html) is its replacement, and I thought we had looked into that when we were considering the JDK9+ port. Maybe @jmagne or @edewata can figure out where we had those discussions back when Dogtag was still Java 8-only.

I think for this change to work, you'd need to run the `wrap(...)` call and store the results of it _before_ you call cleanup.

A counter-test would be to present a bad CA chain as a client. This should trigger a server to verify the chain (during an unwrap) and subsequently need to send an outbound alert (`NEED_WRAP`).

What I expect to happen -- and again, disclaimers about memory aside, happy to be wrong here -- is that because you've cleaned up the SSLFD and underlying NSS sockets, you'd get no data to wrap. This means the client would "just" see a peer disconnect error and not see an actual verification error. 

Anyhow, some thoughts from the peanut gallery :D

Reply:
> Wow, this is a really interesting change.
> 
> I think, `finalize` made sense to me, as a catch-all for any failures of applications to conform with expectations. [If my memory serves correctly, `Cleaner`](https://docs.oracle.com/javase/9/docs/api/java/lang/ref/Cleaner.html) is its replacement, and I thought we had looked into that when we were considering the JDK9+ port. Maybe @jmagne or @edewata can figure out where we had those discussions back when Dogtag was still Java 8-only.
> 

Yes, `Cleaner` is a possible alternative but I do not think it is needed for this case. Nevertheless, I am reverting the `finlize()` for the moment and if it is the case re-evaluate the approach in a future PR.


> I think for this change to work, you'd need to run the `wrap(...)` call and store the results of it _before_ you call cleanup.
> 
> A counter-test would be to present a bad CA chain as a client. This should trigger a server to verify the chain (during an unwrap) and subsequently need to send an outbound alert (`NEED_WRAP`).
> 
> What I expect to happen -- and again, disclaimers about memory aside, happy to be wrong here -- is that because you've cleaned up the SSLFD and underlying NSS sockets, you'd get no data to wrap. This means the client would "just" see a peer disconnect error and not see an actual verification error.
> 

Yes, I think for this you are totally right, so I have reverted the change.

> Anyhow, some thoughts from the peanut gallery :D

Really appreciate, thanks!